File: Add Binary - Easy/addbinary.cpp
#include<bits/stdc++.h>

using namespace std;

string result = "";

void rec(string a, int ai, string b, int bi, bool ex){
    char temp_a = '0';
    char temp_b = '0';
    if (ai < 0 && bi < 0){
        if (ex)
            result = "1" + result;
        return;
    }
    if (ai >= 0)
        temp_a = a[ai];
    if (bi >= 0)
        temp_b = b[bi];
    
    if (temp_a == temp_b && temp_a == '1'){
        result = ((ex) ? '1' : '0') + result;
        ex = true;
    } else if (temp_a == temp_b && temp_a == '0'){
        result = ((ex) ? '1' : '0') + result;
        ex = false;
    } else {
        if (ex){
            result = '0' + result;
            ex = true;
        } else {
            result = '1' + result;
            ex = false;
        }
    }
    rec(a, ai - 1, b, bi - 1, ex);
}

string addBinary(string a, string b) {
    result = "";
    rec(a, a.length() - 1, b, b.length() - 1, false);
    return result;
}

int main(){
	
	return 0;
}

File: Add Binary - Easy/AddBinary.py
result = ""
def rec(a, ai, b, bi, ex):
    global result
    temp_a = "0"
    temp_b = "0"
    if ai < 0 and bi < 0:
        if ex:
            result = "1" + result
        return
    if ai >= 0:
        temp_a = a[ai]
    if bi >= 0:
        temp_b = b[bi]
    if temp_a == temp_b and temp_a == "1":
        result = ("1" if ex else "0") + result
        ex = True
    elif temp_a == temp_b and temp_a == "0":
        result = ("1" if ex else "0") + result
        ex = False
    else:
        if ex:
            result = "0" + result
            ex = True
        else:
            result = "1" + result
            ex = False
    rec(a, ai - 1, b, bi - 1, ex)


def addBinary(a: str, b: str) -> str:
    global result
    result = ""
    rec(a, len(a) - 1, b, len(b) - 1, False)
    return result

# print(addBinary("11", "1"))
# print(addBinary("1010", "1011"))
print(addBinary("1111", "1111"))
